drivers/src/pmu.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2022 by Rivos Inc.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use arrayvec::ArrayVec;
+use sbi::{api::pmu, Error as SbiError, PmuCounterInfo, Result as SbiResult};
+use spin::Once;
+
+/// Maximum number of supported platform PMU counters
+pub const MAX_HARDWARE_COUNTERS: usize = 32;
+/// Caches information about platform hardware counters.
+#[derive(Debug)]
+pub struct PmuInfo {
+    counters_info: ArrayVec<Option<PmuCounterInfo>, MAX_HARDWARE_COUNTERS>,
+    counters_mask: u64,
+}
+
+impl PmuInfo {
+    fn new() -> Self {
+        Self {
+            counters_info: ArrayVec::new(),
+            counters_mask: 0,
+        }
+    }
+
+    fn validate_hardware_counter(info: PmuCounterInfo) -> Option<PmuCounterInfo> {
+        use riscv_regs::{CSR_CYCLE, CSR_HPMCOUNTER31};
+        if info.is_hardware_counter()
+            && info.get_csr() >= (CSR_CYCLE as u64)
+            && info.get_csr() <= (CSR_HPMCOUNTER31 as u64)
+        {
+            Some(info)
+        } else {
+            None
+        }
+    }
+
+    fn init_counter(&mut self, counter_index: u64) -> bool {
+        // Assume that the platform enumerates all HW counters upfront, and ignore firmware counters
+        if let Ok(info) = pmu::get_counter_info(counter_index) {
+            if info.is_firmware_counter() {
+                return false;
+            }
+            let result = self
+                .counters_info
+                .try_push(Self::validate_hardware_counter(info))
```

Comment:
The OpenSBI platform actually doesn't enumerate anything for counter #1 (don't know the historical reasons behind it). It's enumerated to maintain a 1:1 mapping between the counter indexes for the platform guest counter indexes. As I see it, this approach is simpler than trying to perform a mapping (and reverse mapping) of counter indexes (and associated counter masks).

Reply:
counter 1 is returned as invalid because there is no hardware counter 1. This simplifies the mapping.

get_counter_info will return EINVAL for counter 1. So you don't even need to validate _hardware_counter. So a hardware counter with bad CSR value, you can bail out.